crates/bevy_render/src/camera/camera.rs

```diff
@@ -343,19 +347,47 @@ impl Camera {
     /// To get the world space coordinates with the viewport position, you should use
     /// [`world_to_viewport`](Self::world_to_viewport).
     ///
-    /// Returns `None` if the `camera_transform`, the `world_position`, or the projection matrix defined by [`CameraProjection`] contain `NAN`.
     /// Panics if the projection matrix is null and `glam_assert` is enabled.
-    pub fn ndc_to_world(&self, camera_transform: &GlobalTransform, ndc: Vec3) -> Option<Vec3> {
+    pub fn ndc_to_world(
+        &self,
+        camera_transform: &GlobalTransform,
+        ndc: Vec3,
+    ) -> Result<Vec3, NdcToWorldError> {
+        let cam_trans_mat = camera_transform.compute_matrix();
+        if cam_trans_mat.is_nan() {
+            return Err(NdcToWorldError::CameraTransformNanError(cam_trans_mat));
+        }
+
+        let proj_mat = self.computed.projection_matrix;
+        if proj_mat.is_nan() {
+            return Err(NdcToWorldError::ProjectionMatrixNanError(proj_mat));
+        }
+
         // Build a transformation matrix to convert from NDC to world space using camera data
-        let ndc_to_world =
-            camera_transform.compute_matrix() * self.computed.projection_matrix.inverse();
+        let ndc_to_world = cam_trans_mat * proj_mat.inverse();
 
         let world_space_coords = ndc_to_world.project_point3(ndc);
+        if world_space_coords.is_nan() {
+            return Err(NdcToWorldError::WorldSpaceCoordsNanError(
+                world_space_coords,
+            ));
+        }
 
-        (!world_space_coords.is_nan()).then_some(world_space_coords)
+        Ok(world_space_coords)
     }
 }
 
+pub enum NdcToWorldError {
+    WorldSpaceCoordsNanError(Vec3),
+    CameraTransformNanError(Mat4),
+    ProjectionMatrixNanError(Mat4),
+}
+
+pub enum ViewportToWorldError {
```

Comment:
I guess it's fine to reuse this enum for the 2d variant, but the 3d variant should definitely use it.

Reply:
Didn't quite understand this. Is there a "not" missing from that sentence somewhere?

Reply:
No, my point is that the PR currently only uses this enum for viewport_to_world_2d() but it should also be ysed for viewport_to_world(). I just didn't realize this was a draft PR.